crates/dojo-defi/src/constant_product_market/components.cairo

```diff
@@ -40,6 +79,189 @@ impl MarketImpl of MarketTrait {
+    }
+
+    // Get the liquidity of the market
+    // Use cubit fixed point math library to compute the square root of the product of the reserves
+    fn liquidity(self: @Market) -> FixedType {
+        // Get normalized reserve cash amount and item quantity
+        let (reserve_amount, reserve_quantity) = self.get_reserves();
+
+        // Convert reserve amount to fixed point
+        let reserve_amount = Fixed::new_unscaled(reserve_amount, false);
+        let reserve_quantity = Fixed::new_unscaled(reserve_quantity, false);
+
+        // L = sqrt(X * Y)
+        (reserve_amount * reserve_quantity).sqrt()
+    }
+
+    // Check if the market has liquidity
+    fn has_liquidity(self: @Market) -> bool {
+        *self.cash_amount > 0 | *self.item_quantity > 0
+    }
+
+    // Given some amount of cash, return the equivalent/optimal quantity of items
+    // based on the reserves in the market
+    fn quote_quantity(self: @Market, amount: u128) -> usize {
+        assert(amount > 0, 'insufficient amount');
+        assert(self.has_liquidity(), 'insufficient liquidity');
+
+        // Get normalized reserve cash amount and item quantity
+        let (reserve_amount, reserve_quantity) = self.get_reserves();
+
+        // Convert amount to fixed point
+        let amount = Fixed::new_unscaled(amount, false);
+
+        // Convert reserve amount and quantity to fixed point
+        let reserve_amount = Fixed::new_unscaled(reserve_amount, false);
+        let reserve_quantity = Fixed::new_unscaled(reserve_quantity, false);
+
+        // dy = Y * dx / X
+        let quantity_optimal = (reserve_quantity * amount) / reserve_amount;
+        // Unscale and convert to usize
+        (quantity_optimal.into().try_into().unwrap() / ONE_u128).into().try_into().unwrap()
+    }
+
+    // Given some quantity of items, return the equivalent/optimal amount of cash
+    // based on the reserves in the market
+    fn quote_amount(self: @Market, quantity: usize) -> u128 {
+        assert(quantity > 0, 'insufficient quantity');
+        assert(self.has_liquidity(), 'insufficient liquidity');
+
+        // Get normalized reserve cash amount and item quantity
+        let (reserve_amount, reserve_quantity) = self.get_reserves();
+
+        // Convert reserve amount and quantity to fixed point
+        let reserve_amount = Fixed::new_unscaled(reserve_amount, false);
+        let reserve_quantity = Fixed::new_unscaled(reserve_quantity, false);
+
+        // Normalize quantity
+        let quantity = quantity.into().try_into().unwrap() * SCALING_FACTOR;
+
+        // Convert quantity to fixed point
+        let quantity = Fixed::new_unscaled(quantity, false);
+
+        // dx = X * dy / Y
+        let amount_optimal = (reserve_amount * quantity) / reserve_quantity;
+        // Convert amount to u128 and unscale
+        amount_optimal.into().try_into().unwrap() / ONE_u128
+    }
```

Comment:
This implementation follows the one from [UniswapV2Router02](https://etherscan.io/address/0x7a250d5630b4cf539739df2c5dacb4c659f2488d). Gets the quote amount/quantity that is equivalent to the given quantity/amount provided so that when liquidity is added or removed, the ratio doesn't change